core/amber/src/main/scala/edu/uci/ics/amber/engine/common/promise/RPCServer.scala

```diff
@@ -0,0 +1,53 @@
+package edu.uci.ics.amber.engine.common.promise
+
+import com.twitter.util.Future
+import edu.uci.ics.amber.engine.architecture.messaginglayer.ControlOutputPort
+import edu.uci.ics.amber.engine.common.ambertag.neo.VirtualIdentity.ActorVirtualIdentity
+import edu.uci.ics.amber.engine.common.promise.RPCClient.{ControlInvocation, ReturnPayload}
+import edu.uci.ics.amber.engine.common.promise.RPCServer.RPCCommand
+
+object RPCServer {
+
+  final val RootCallWithoutReturn = -1
+
+  trait RPCCommand[T]
+}
+
+class RPCServer(controlOutputPort: ControlOutputPort) {
+
+  // all handlers
+  protected var handlers: PartialFunction[RPCCommand[_], Any] = PartialFunction.empty
+
+  def registerHandler(newHandler: PartialFunction[RPCCommand[_], Any]): Unit = {
+    handlers =
+      newHandler orElse handlers
```

Comment:
This doesnt prohibit two handlers for the same control message

Reply:
Yes, but the latter one will be used. It's consistent as using a map without check if it contains the key. I think there is no way to check if a partial function contains a certain case. And I didn't find another way to map types to lambdas.

Reply:
We can only prevent this by not writing 2 handlers for the same control.

Reply:
That's fine. Just add a comment above clarifying this.